File: src/main/java/net/minestom/server/listener/manager/PacketListenerManager.java
package net.minestom.server.listener.manager;

import net.minestom.server.MinecraftServer;
import net.minestom.server.event.EventDispatcher;
import net.minestom.server.event.player.PlayerPacketEvent;
import net.minestom.server.listener.*;
import net.minestom.server.listener.common.KeepAliveListener;
import net.minestom.server.listener.common.PluginMessageListener;
import net.minestom.server.listener.common.ResourcePackListener;
import net.minestom.server.listener.common.SettingsListener;
import net.minestom.server.listener.preplay.ConfigListener;
import net.minestom.server.listener.preplay.HandshakeListener;
import net.minestom.server.listener.preplay.LoginListener;
import net.minestom.server.listener.preplay.StatusListener;
import net.minestom.server.network.ConnectionState;
import net.minestom.server.network.packet.client.ClientPacket;
import net.minestom.server.network.packet.client.common.*;
import net.minestom.server.network.packet.client.configuration.ClientFinishConfigurationPacket;
import net.minestom.server.network.packet.client.handshake.ClientHandshakePacket;
import net.minestom.server.network.packet.client.login.ClientEncryptionResponsePacket;
import net.minestom.server.network.packet.client.login.ClientLoginAcknowledgedPacket;
import net.minestom.server.network.packet.client.login.ClientLoginPluginResponsePacket;
import net.minestom.server.network.packet.client.login.ClientLoginStartPacket;
import net.minestom.server.network.packet.client.play.*;
import net.minestom.server.network.packet.client.status.StatusRequestPacket;
import net.minestom.server.network.player.PlayerConnection;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public final class PacketListenerManager {

    private final static Logger LOGGER = LoggerFactory.getLogger(PacketListenerManager.class);

    private final Map<Class<? extends ClientPacket>, PacketPrePlayListenerConsumer>[] listeners = new Map[ConnectionState.values().length];

    public PacketListenerManager() {
        for (int i = 0; i < listeners.length; i++) {
            listeners[i] = new ConcurrentHashMap<>();
        }

        setListener(ConnectionState.HANDSHAKE, ClientHandshakePacket.class, HandshakeListener::listener);

        setListener(ConnectionState.STATUS, StatusRequestPacket.class, StatusListener::requestListener);
        setListener(ConnectionState.STATUS, ClientPingRequestPacket.class, StatusListener::pingRequestListener);

        setListener(ConnectionState.LOGIN, ClientLoginStartPacket.class, LoginListener::loginStartListener);
        setListener(ConnectionState.LOGIN, ClientEncryptionResponsePacket.class, LoginListener::loginEncryptionResponseListener);
        setListener(ConnectionState.LOGIN, ClientLoginPluginResponsePacket.class, LoginListener::loginPluginResponseListener);
        setListener(ConnectionState.LOGIN, ClientLoginAcknowledgedPacket.class, LoginListener::loginAckListener);

        setConfigurationListener(ClientSettingsPacket.class, SettingsListener::listener);
        setConfigurationListener(ClientPluginMessagePacket.class, PluginMessageListener::listener);
        setConfigurationListener(ClientKeepAlivePacket.class, KeepAliveListener::listener);
        setConfigurationListener(ClientPongPacket.class, (packet, player) -> {/* empty */});
        setConfigurationListener(ClientResourcePackStatusPacket.class, ResourcePackListener::listener);
        setConfigurationListener(ClientFinishConfigurationPacket.class, ConfigListener::finishConfigListener);

        setPlayListener(ClientKeepAlivePacket.class, KeepAliveListener::listener);
        setPlayListener(ClientCommandChatPacket.class, ChatMessageListener::commandChatListener);
        setPlayListener(ClientChatMessagePacket.class, ChatMessageListener::chatMessageListener);
        setPlayListener(ClientClickWindowPacket.class, WindowListener::clickWindowListener);
        setPlayListener(ClientCloseWindowPacket.class, WindowListener::closeWindowListener);
        setPlayListener(ClientClickWindowButtonPacket.class, WindowListener::buttonClickListener);
        setPlayListener(ClientConfigurationAckPacket.class, PlayConfigListener::configAckListener);
        setPlayListener(ClientPongPacket.class, WindowListener::pong);
        setPlayListener(ClientEntityActionPacket.class, EntityActionListener::listener);
        setPlayListener(ClientHeldItemChangePacket.class, PlayerHeldListener::heldListener);
        setPlayListener(ClientPlayerBlockPlacementPacket.class, BlockPlacementListener::listener);
        setPlayListener(ClientSteerVehiclePacket.class, PlayerVehicleListener::steerVehicleListener);
        setPlayListener(ClientVehicleMovePacket.class, PlayerVehicleListener::vehicleMoveListener);
        setPlayListener(ClientSteerBoatPacket.class, PlayerVehicleListener::boatSteerListener);
        setPlayListener(ClientPlayerPacket.class, PlayerPositionListener::playerPacketListener);
        setPlayListener(ClientPlayerRotationPacket.class, PlayerPositionListener::playerLookListener);
        setPlayListener(ClientPlayerPositionPacket.class, PlayerPositionListener::playerPositionListener);
        setPlayListener(ClientPlayerPositionAndRotationPacket.class, PlayerPositionListener::playerPositionAndLookListener);
        setPlayListener(ClientTeleportConfirmPacket.class, PlayerPositionListener::teleportConfirmListener);
        setPlayListener(ClientPlayerDiggingPacket.class, PlayerDiggingListener::playerDiggingListener);
        setPlayListener(ClientAnimationPacket.class, AnimationListener::animationListener);
        setPlayListener(ClientInteractEntityPacket.class, UseEntityListener::useEntityListener);
        setPlayListener(ClientUseItemPacket.class, UseItemListener::useItemListener);
        setPlayListener(ClientStatusPacket.class, PlayStatusListener::listener);
        setPlayListener(ClientSettingsPacket.class, SettingsListener::listener);
        setPlayListener(ClientCreativeInventoryActionPacket.class, CreativeInventoryActionListener::listener);
        setPlayListener(ClientCraftRecipeRequest.class, RecipeListener::listener);
        setPlayListener(ClientTabCompletePacket.class, TabCompleteListener::listener);
        setPlayListener(ClientPluginMessagePacket.class, PluginMessageListener::listener);
        setPlayListener(ClientPlayerAbilitiesPacket.class, AbilitiesListener::listener);
        setPlayListener(ClientResourcePackStatusPacket.class, ResourcePackListener::listener);
        setPlayListener(ClientAdvancementTabPacket.class, AdvancementTabListener::listener);
        setPlayListener(ClientSpectatePacket.class, SpectateListener::listener);
        setPlayListener(ClientEditBookPacket.class, BookListener::listener);
        setPlayListener(ClientChatSessionUpdatePacket.class, (packet, player) -> {/* empty */});
        setPlayListener(ClientChunkBatchReceivedPacket.class, ChunkBatchListener::batchReceivedListener);
        setPlayListener(ClientPingRequestPacket.class, PlayPingListener::requestListener);
    }

    /**
     * Processes a packet by getting its {@link PacketPlayListenerConsumer} and calling all the packet listeners.
     *
     * @param packet     the received packet
     * @param connection the connection of the player who sent the packet
     * @param <T>        the packet type
     */
    public <T extends ClientPacket> void processClientPacket(@NotNull T packet, @NotNull PlayerConnection connection) {
        final ConnectionState state = connection.getConnectionState();
        final Class clazz = packet.getClass();
        PacketPrePlayListenerConsumer<T> packetListenerConsumer = listeners[state.ordinal()].get(clazz);

        // Listener can be null if none has been set before, call PacketConsumer anyway
        if (packetListenerConsumer == null) {
            LOGGER.warn("Packet " + clazz + " does not have any default listener! (The issue comes from Minestom)");
            return;
        }

        // Event
        if (state == ConnectionState.PLAY) {
            PlayerPacketEvent playerPacketEvent = new PlayerPacketEvent(connection.getPlayer(), packet);
            EventDispatcher.call(playerPacketEvent);
            if (playerPacketEvent.isCancelled()) {
                return;
            }
        }

        // Finally execute the listener
        try {
            packetListenerConsumer.accept(packet, connection);
        } catch (Exception e) {
            // Packet is likely invalid
            MinecraftServer.getExceptionManager().handleException(e);
        }
    }

    /**
     * Sets the listener of a packet.
     * <p>
     * WARNING: this will overwrite the default minestom listener, this is not reversible.
     *
     * @param state       the state of the packet
     * @param packetClass the class of the packet
     * @param consumer    the new packet's listener
     * @param <T>         the type of the packet
     */
    public <T extends ClientPacket> void setListener(@NotNull ConnectionState state, @NotNull Class<T> packetClass, @NotNull PacketPrePlayListenerConsumer<T> consumer) {
        this.listeners[state.ordinal()].put(packetClass, consumer);
    }

    /**
     * Sets the listener of a packet.
     * <p>
     * WARNING: this will overwrite the default minestom listener, this is not reversible.
     *
     * @param packetClass the class of the packet
     * @param consumer    the new packet's listener
     * @param <T>         the type of the packet
     */
    public <T extends ClientPacket> void setPlayListener(@NotNull Class<T> packetClass, @NotNull PacketPlayListenerConsumer<T> consumer) {
        setListener(ConnectionState.PLAY, packetClass, (packet, playerConnection) -> consumer.accept(packet, playerConnection.getPlayer()));
    }

    public <T extends ClientPacket> void setConfigurationListener(@NotNull Class<T> packetClass, @NotNull PacketPlayListenerConsumer<T> consumer) {
        setListener(ConnectionState.CONFIGURATION, packetClass, (packet, playerConnection) -> consumer.accept(packet, playerConnection.getPlayer()));
    }

    /**
     * Sets the listener of a packet.
     * <p>
     * WARNING: this will overwrite the default minestom listener, this is not reversible.
     *
     * @param packetClass the class of the packet
     * @param consumer    the new packet's listener
     * @param <T>         the type of the packet
     */
    @Deprecated
    public <T extends ClientPacket> void setListener(@NotNull Class<T> packetClass, @NotNull PacketPlayListenerConsumer<T> consumer) {
        setPlayListener(packetClass, consumer);
    }

}

File: src/main/java/net/minestom/server/listener/manager/PacketListenerManager.kt
package net.minestom.server.listener.manager

import net.minestom.server.MinecraftServer.Companion.exceptionManager
import net.minestom.server.ServerProcess
import net.minestom.server.network.packet.client.ClientPacket
import net.minestom.server.listener.manager.PacketListenerConsumer
import java.util.concurrent.ConcurrentHashMap
import net.minestom.server.network.packet.client.play.ClientKeepAlivePacket
import net.minestom.server.network.packet.client.play.ClientChatMessagePacket
import net.minestom.server.network.packet.client.play.ClientClickWindowPacket
import net.minestom.server.network.packet.client.play.ClientCloseWindowPacket
import net.minestom.server.network.packet.client.play.ClientPongPacket
import net.minestom.server.network.packet.client.play.ClientEntityActionPacket
import net.minestom.server.network.packet.client.play.ClientHeldItemChangePacket
import net.minestom.server.network.packet.client.play.ClientPlayerBlockPlacementPacket
import net.minestom.server.network.packet.client.play.ClientSteerVehiclePacket
import net.minestom.server.network.packet.client.play.ClientVehicleMovePacket
import net.minestom.server.network.packet.client.play.ClientSteerBoatPacket
import net.minestom.server.network.packet.client.play.ClientPlayerPacket
import net.minestom.server.network.packet.client.play.ClientPlayerRotationPacket
import net.minestom.server.network.packet.client.play.ClientPlayerPositionPacket
import net.minestom.server.network.packet.client.play.ClientPlayerPositionAndRotationPacket
import net.minestom.server.network.packet.client.play.ClientTeleportConfirmPacket
import net.minestom.server.network.packet.client.play.ClientPlayerDiggingPacket
import net.minestom.server.network.packet.client.play.ClientAnimationPacket
import net.minestom.server.network.packet.client.play.ClientInteractEntityPacket
import net.minestom.server.network.packet.client.play.ClientUseItemPacket
import net.minestom.server.network.packet.client.play.ClientStatusPacket
import net.minestom.server.network.packet.client.play.ClientSettingsPacket
import net.minestom.server.network.packet.client.play.ClientCreativeInventoryActionPacket
import net.minestom.server.network.packet.client.play.ClientCraftRecipeRequest
import net.minestom.server.network.packet.client.play.ClientTabCompletePacket
import net.minestom.server.network.packet.client.play.ClientPluginMessagePacket
import net.minestom.server.network.packet.client.play.ClientPlayerAbilitiesPacket
import net.minestom.server.network.packet.client.play.ClientResourcePackStatusPacket
import net.minestom.server.network.packet.client.play.ClientAdvancementTabPacket
import net.minestom.server.network.packet.client.play.ClientSpectatePacket
import net.minestom.server.listener.manager.PacketListenerManager
import net.minestom.server.event.player.PlayerPacketEvent
import net.minestom.server.MinecraftServer
import net.minestom.server.entity.Player
import net.minestom.server.event.EventDispatcher
import net.minestom.server.listener.*
import org.slf4j.LoggerFactory
import java.lang.Exception

class PacketListenerManager(private val serverProcess: ServerProcess) {
    private val listeners: MutableMap<Class<out ClientPacket>, PacketListenerConsumer<*>> = ConcurrentHashMap()

    init {
        setListener(
            ClientKeepAlivePacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> KeepAliveListener.listener(packet, player) })
        setListener(
            ClientChatMessagePacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> ChatMessageListener.listener(packet, player) })
        setListener(
            ClientClickWindowPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                WindowListener.clickWindowListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientCloseWindowPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                WindowListener.closeWindowListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientPongPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> WindowListener.pong(packet, player) })
        setListener(
            ClientEntityActionPacket::class.java,
            PacketListenerConsumer { obj: T?, packet: Player? -> obj.listener(packet) })
        setListener(
            ClientHeldItemChangePacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> PlayerHeldListener.heldListener(packet, player) })
        setListener(
            ClientPlayerBlockPlacementPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> BlockPlacementListener.listener(packet, player) })
        setListener(
            ClientSteerVehiclePacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                PlayerVehicleListener.steerVehicleListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientVehicleMovePacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                PlayerVehicleListener.vehicleMoveListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientSteerBoatPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                PlayerVehicleListener.boatSteerListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientPlayerPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                PlayerPositionListener.playerPacketListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientPlayerRotationPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                PlayerPositionListener.playerLookListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientPlayerPositionPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                PlayerPositionListener.playerPositionListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientPlayerPositionAndRotationPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                PlayerPositionListener.playerPositionAndLookListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientTeleportConfirmPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                PlayerPositionListener.teleportConfirmListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientPlayerDiggingPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                PlayerDiggingListener.playerDiggingListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientAnimationPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                AnimationListener.animationListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientInteractEntityPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                UseEntityListener.useEntityListener(
                    packet,
                    player
                )
            })
        setListener(
            ClientUseItemPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> UseItemListener.useItemListener(packet, player) })
        setListener(
            ClientStatusPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> StatusListener.listener(packet, player) })
        setListener(
            ClientSettingsPacket::class.java,
            PacketListenerConsumer { obj: T?, packet: Player? -> obj.listener(packet) })
        setListener(
            ClientCreativeInventoryActionPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? ->
                CreativeInventoryActionListener.listener(
                    packet,
                    player
                )
            })
        setListener(
            ClientCraftRecipeRequest::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> RecipeListener.listener(packet, player) })
        setListener(
            ClientTabCompletePacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> TabCompleteListener.listener(packet, player) })
        setListener(
            ClientPluginMessagePacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> PluginMessageListener.listener(packet, player) })
        setListener(
            ClientPlayerAbilitiesPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> AbilitiesListener.listener(packet, player) })
        setListener(
            ClientResourcePackStatusPacket::class.java,
            PacketListenerConsumer { obj: T?, packet: Player? -> obj.listener(packet) })
        setListener(
            ClientAdvancementTabPacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> AdvancementTabListener.listener(packet, player) })
        setListener(
            ClientSpectatePacket::class.java,
            PacketListenerConsumer { packet: T?, player: Player? -> SpectateListener.listener(packet, player) })
    }

    /**
     * Processes a packet by getting its [PacketListenerConsumer] and calling all the packet listeners.
     *
     * @param packet the received packet
     * @param player the player who sent the packet
     * @param <T>    the packet type
    </T> */
    fun <T : ClientPacket?> processClientPacket(packet: T, player: Player) {
        val clazz: Class<*> = packet.javaClass
        val packetListenerConsumer: PacketListenerConsumer<T>? = listeners.get(clazz)

        // Listener can be null if none has been set before, call PacketConsumer anyway
        if (packetListenerConsumer == null) {
            LOGGER.warn("Packet $clazz does not have any default listener! (The issue comes from Minestom)")
        }

        // Event
        val playerPacketEvent = PlayerPacketEvent(player, packet)
        EventDispatcher.call(playerPacketEvent)
        if (playerPacketEvent.isCancelled) {
            return
        }

        // Finally execute the listener
        if (packetListenerConsumer != null) {
            try {
                packetListenerConsumer.accept(packet, player)
            } catch (e: Exception) {
                // Packet is likely invalid
                exceptionManager.handleException(e)
            }
        }
    }

    /**
     * Sets the listener of a packet.
     *
     *
     * WARNING: this will overwrite the default minestom listener, this is not reversible.
     *
     * @param packetClass the class of the packet
     * @param consumer    the new packet's listener
     * @param <T>         the type of the packet
    </T> */
    fun <T : ClientPacket?> setListener(packetClass: Class<T>, consumer: PacketListenerConsumer<T>) {
        listeners[packetClass] = consumer
    }

    companion object {
        val LOGGER = LoggerFactory.getLogger(PacketListenerManager::class.java)
    }
}